### ISSUE TITLE
aws_db_security_group: terraform crash and incorrect documentation

### ISSUE THREAD

**Issue by kestred** [hashicorp/terraform#4352]
Using terraform 0.6.8, given an aws_db_security_group like the following:

```
resource "aws_db_security_group" "example" {
     name = "example database"
     description = "example description"
    ingress {
    }
 }
 ```

An apply results in `!!!!!!!!!!!!!!!!!!!!!!!!!!! TERRAFORM CRASH !!!!!!!!!!!!!!!!!!!!!!!!!!!!`

I'm guessing this occurs because the ingress is empty.
**ASIDE:** The [documentation](https://www.terraform.io/docs/providers/aws/r/db_security_group.html) lists **ingress** as optional, but terraform 0.6.8 considers it required.

See stacktrace:
```
2015/12/16 14:21:24 [ERR] yamux: Failed to write header: write unix @->/tmp/tf-plugin392358876: write: broken pipe
2015/12/16 14:21:24 [DEBUG] terraform-provider-aws: panic: interface conversion: interface {} is nil, not map[string]interface {}
2015/12/16 14:21:24 [DEBUG] terraform-provider-aws: 
2015/12/16 14:21:24 [DEBUG] terraform-provider-aws: goroutine 346 [running]:
2015/12/16 14:21:24 [DEBUG] terraform-provider-aws: github.com/hashicorp/terraform/builtin/providers/aws.resourceAwsDbSecurityGroupIngressHash(0x0, 0x0, 0xc800000001)
2015/12/16 14:21:24 [DEBUG] terraform-provider-aws: 	/opt/gopath/src/github.com/hashicorp/terraform/builtin/providers/aws/resource_aws_db_security_group.go:244 +0x8ec
2015/12/16 14:21:24 [DEBUG] terraform-provider-aws: github.com/hashicorp/terraform/helper/schema.(*Set).hash(0xc8203b5a60, 0x0, 0x0, 0xc8204b4f70)
2015/12/16 14:21:24 [DEBUG] terraform-provider-aws: 	/opt/gopath/src/github.com/hashicorp/terraform/helper/schema/set.go:175 +0x34
2015/12/16 14:21:24 [DEBUG] terraform-provider-aws: github.com/hashicorp/terraform/helper/schema.(*Set).add(0xc8203b5a60, 0x0, 0x0, 0xc8204b5520)
2015/12/16 14:21:24 [DEBUG] terraform-provider-aws: 	/opt/gopath/src/github.com/hashicorp/terraform/helper/schema/set.go:166 +0x6f
2015/12/16 14:21:24 [DEBUG] terraform-provider-aws: github.com/hashicorp/terraform/helper/schema.(*ConfigFieldReader).readSet(0xc8203b5940, 0xc8203b2980, 0x1, 0x1, 0xc8204b5520, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
2015/12/16 14:21:24 [DEBUG] terraform-provider-aws: 	/opt/gopath/src/github.com/hashicorp/terraform/helper/schema/field_reader_config.go:229 +0x678
2015/12/16 14:21:24 [DEBUG] terraform-provider-aws: github.com/hashicorp/terraform/helper/schema.(*ConfigFieldReader).readField(0xc8203b5940, 0xc8203b2980, 0x1, 0x1, 0xc8203b5a00, 0x0, 0x0, 0x0, 0x0, 0xc820330000, ...)
2015/12/16 14:21:24 [DEBUG] terraform-provider-aws: 	/opt/gopath/src/github.com/hashicorp/terraform/helper/schema/field_reader_config.go:90 +0x6d9
2015/12/16 14:21:24 [DEBUG] terraform-provider-aws: github.com/hashicorp/terraform/helper/schema.(*ConfigFieldReader).ReadField(0xc8203b5940, 0xc8203b2980, 0x1, 0x1, 0x0, 0x0, 0x0, 0x0, 0x640000, 0x0, ...)
2015/12/16 14:21:24 [DEBUG] terraform-provider-aws: 	/opt/gopath/src/github.com/hashicorp/terraform/helper/schema/field_reader_config.go:25 +0x8d
2015/12/16 14:21:24 [DEBUG] terraform-provider-aws: github.com/hashicorp/terraform/helper/schema.(*MultiLevelFieldReader).ReadFieldExact(0xc8203b5980, 0xc8203b2980, 0x1, 0x1, 0x118d4e8, 0x6, 0x0, 0x0, 0x0, 0x0, ...)
2015/12/16 14:21:24 [DEBUG] terraform-provider-aws: 	/opt/gopath/src/github.com/hashicorp/terraform/helper/schema/field_reader_multi.go:31 +0x27c
2015/12/16 14:21:24 [DEBUG] terraform-provider-aws: github.com/hashicorp/terraform/helper/schema.(*ResourceData).get(0xc8202aad80, 0xc8203b2980, 0x1, 0x1, 0xc8203b2902, 0x0, 0x0, 0x0, 0x0, 0x400000, ...)
2015/12/16 14:21:24 [DEBUG] terraform-provider-aws: 	/opt/gopath/src/github.com/hashicorp/terraform/helper/schema/resource_data.go:387 +0x289
2015/12/16 14:21:24 [DEBUG] terraform-provider-aws: github.com/hashicorp/terraform/helper/schema.(*ResourceData).getChange(0xc8202aad80, 0x118ec40, 0x7, 0x641201, 0x0, 0x0, 0x0, 0x0, 0xc8204b0000, 0x0, ...)
2015/12/16 14:21:24 [DEBUG] terraform-provider-aws: 	/opt/gopath/src/github.com/hashicorp/terraform/helper/schema/resource_data.go:356 +0x235
2015/12/16 14:21:24 [DEBUG] terraform-provider-aws: github.com/hashicorp/terraform/helper/schema.(*ResourceData).diffChange(0xc8202aad80, 0x118ec40, 0x7, 0x0, 0x0, 0x0, 0x0, 0x1)
2015/12/16 14:21:24 [DEBUG] terraform-provider-aws: 	/opt/gopath/src/github.com/hashicorp/terraform/helper/schema/resource_data.go:333 +0x78
2015/12/16 14:21:24 [DEBUG] terraform-provider-aws: github.com/hashicorp/terraform/helper/schema.schemaMap.diffSet(0xc820589830, 0x118ec40, 0x7, 0xc8204b5520, 0xc8203b2770, 0xc8202aad80, 0x0, 0x0, 0x0)
2015/12/16 14:21:24 [DEBUG] terraform-provider-aws: 	/opt/gopath/src/github.com/hashicorp/terraform/helper/schema/schema.go:793 +0x7b
2015/12/16 14:21:24 [DEBUG] terraform-provider-aws: github.com/hashicorp/terraform/helper/schema.schemaMap.diff(0xc820589830, 0x118ec40, 0x7, 0xc8204b5520, 0xc8203b2770, 0xc8202aad80, 0xc81ffe2000, 0x0, 0x0)
2015/12/16 14:21:24 [DEBUG] terraform-provider-aws: 	/opt/gopath/src/github.com/hashicorp/terraform/helper/schema/schema.go:567 +0x32e
2015/12/16 14:21:24 [DEBUG] terraform-provider-aws: github.com/hashicorp/terraform/helper/schema.schemaMap.Diff(0xc820589830, 0xc82024f290, 0xc82024f350, 0xc8203a8950, 0x0, 0x0)
2015/12/16 14:21:24 [DEBUG] terraform-provider-aws: 	/opt/gopath/src/github.com/hashicorp/terraform/helper/schema/schema.go:296 +0x1f5
2015/12/16 14:21:24 [DEBUG] terraform-provider-aws: github.com/hashicorp/terraform/helper/schema.(*Resource).Diff(0xc820384500, 0xc82024f290, 0xc82024f350, 0x15, 0x0, 0x0)
2015/12/16 14:21:24 [DEBUG] terraform-provider-aws: 	/opt/gopath/src/github.com/hashicorp/terraform/helper/schema/resource.go:162 +0x44
2015/12/16 14:21:24 [DEBUG] terraform-provider-aws: github.com/hashicorp/terraform/helper/schema.(*Provider).Diff(0xc8204f6a50, 0xc820265e40, 0xc82024f290, 0xc82024f350, 0x1, 0x0, 0x0)
2015/12/16 14:21:24 [DEBUG] terraform-provider-aws: 	/opt/gopath/src/github.com/hashicorp/terraform/helper/schema/provider.go:175 +0x1c5
2015/12/16 14:21:24 [DEBUG] terraform-provider-aws: github.com/hashicorp/terraform/rpc.(*ResourceProviderServer).Diff(0xc8203e18a0, 0xc8203b51c0, 0xc8203b2760, 0x0, 0x0)
2015/12/16 14:21:24 [DEBUG] terraform-provider-aws: 	/opt/gopath/src/github.com/hashicorp/terraform/rpc/resource_provider.go:334 +0x76
2015/12/16 14:21:24 [DEBUG] terraform-provider-aws: reflect.Value.call(0xef1c00, 0x10f91c0, 0x13, 0x1189660, 0x4, 0xc82061ded8, 0x3, 0x3, 0x0, 0x0, ...)
2015/12/16 14:21:24 [DEBUG] terraform-provider-aws: 	/usr/local/go/src/reflect/value.go:432 +0x120a
2015/12/16 14:21:24 [DEBUG] terraform-provider-aws: reflect.Value.Call(0xef1c00, 0x10f91c0, 0x13, 0xc82061ded8, 0x3, 0x3, 0x0, 0x0, 0x0)
2015/12/16 14:21:24 [DEBUG] terraform-provider-aws: 	/usr/local/go/src/reflect/value.go:300 +0xb1
2015/12/16 14:21:24 [DEBUG] terraform-provider-aws: net/rpc.(*service).call(0xc82031e580, 0xc82031e540, 0xc8203d0b80, 0xc820593700, 0xc8203c2680, 0xcb7520, 0xc8203b51c0, 0x16, 0xcb7580, 0xc8203b2760, ...)
2015/12/16 14:21:24 [DEBUG] terraform-provider-aws: 	/opt/go/src/net/rpc/server.go:383 +0x1c1
2015/12/16 14:21:24 [DEBUG] terraform-provider-aws: created by net/rpc.(*Server).ServeCodec
2015/12/16 14:21:24 [DEBUG] terraform-provider-aws: 	/opt/go/src/net/rpc/server.go:477 +0x4ac
```

**Comment by stack72** [comment]
@kestred as it is optional, you don't even need to specify it. Your code would need to be:

```
resource "aws_db_security_group" "example" {
     name = "example database"
     description = "example description"
 }
```

**Comment by kestred** [comment]
@stack72 it is documented as optional, but terraform complains if it missing saying is required.
Your example does not run correctly when I tried using Terraform (linux, 0.6.8, amd64)

I switched to using the RDS in VPC which is what Amazon prefers going forward anyways which means that aws_db_security_group isn't needed (you use aws_security_group instead).

However, it is still incorrect behavior for terraform to crash while running.
